// nestjs-rmq/dist/helpers/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RQMColorLogger = void 0;
const common_1 = require("@nestjs/common");
const chalk_1 = require("chalk");
class RQMColorLogger {
    constructor(logMessages) {
        this.logMessages = logMessages !== null && logMessages !== void 0 ? logMessages : false;
    }
    log(message, context) {
        common_1.Logger.log(message, context);
    }
    error(message, trace, context) {
        common_1.Logger.error(message, trace, context);
    }
    debug(message, context) {
        if (!this.logMessages) {
            return;
        }
        const msg = JSON.stringify(message);
        const action = context.split(',')[0];
        const topic = context.split(',')[1];
        common_1.Logger.log(`${(0, chalk_1.blueBright)(action)} [${(0, chalk_1.yellow)(topic)}] ${(0, chalk_1.white)(msg)}`);
        console.warn(`${(0, chalk_1.blueBright)(action)} [${(0, chalk_1.yellow)(topic)}] ${(0, chalk_1.white)(msg)}`);
    }
    warn(message, context) {
        common_1.Logger.warn(message, context);
    }
}
exports.RQMColorLogger = RQMColorLogger;
